### PR TITLE
NIFI-3721 Added documentation for Encrypted Provenance Repositories t…

### PR DESCRIPTION
**Pull request by alopresto**
…o Admin Guide and User Guide.

Added screenshot of encrypted provenance repository contents on disk.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by andrewmlim** [comment]
Reviewing

File: nifi-docs/src/main/asciidoc/administration-guide.adoc [hunk 1]
```diff
@@ -2100,7 +2100,7 @@ this property specifies the maximum amount of time to keep the archived data. It
 
 === Provenance Repository
 
-The Provenance Repository contains the information related to Data Provenance. The next three sections are for Provenance Repository properties.
+The Provenance Repository contains the information related to Data Provenance. The next four sections are for Provenance Repository properties.
 
 |====
 |*Property*|*Description*
```
Comment (andrewmlim) [inline nifi-docs/src/main/asciidoc/administration-guide.adoc]: should add formatting to the the values in this first paragraph for consistency ie. (`org.apache.nifi.provenance.PersistentProvenanceRepository` and `org.apache.nifi.provenance.VolatileProvenanceRepository`)

Comment (alopresto) [inline nifi-docs/src/main/asciidoc/administration-guide.adoc]: Will do. Thanks.

File: nifi-docs/src/main/asciidoc/administration-guide.adoc [hunk 2]
```diff
@@ -2110,18 +2110,19 @@ To store provenance events in memory instead of on disk (in which case all event
...
+This implementation was created to replace the `PersistentProvenanceRepository`. The `PersistentProvenanceRepository` was originally written with the simple goal of persisting
 Provenance Events as they are generated and providing the ability to iterate over those events sequentially. Later, it was desired to be able to compress the data so that
 more data could be stored. After that, the ability to index and query the data was added. As requirements evolved over time, the repository kept changing without any major
-redesigns. When used in a NiFi instance that is responsible for processing large volumes of small FlowFiles, the PersistentProvenanceRepository can quickly become a bottleneck.
-The WriteAheadProvenanceRepository was then written to provide the same capabilities as the PersistentProvenanceRepository while providing far better performance.
-Changing to the WriteAheadProvenanceRepository is easy to accomplish, as the two repositories support most of the same properties.
-*Note Well*, however, the follow caveat: The WriteAheadProvenanceRepository will make use of the Provenance data stored by the PersistentProvenanceRepository. However, the
```
Comment (andrewmlim) [inline nifi-docs/src/main/asciidoc/administration-guide.adoc]: change "follow" to "following"

Comment (alopresto) [inline nifi-docs/src/main/asciidoc/administration-guide.adoc]: Will do. Thanks.

File: nifi-docs/src/main/asciidoc/user-guide.adoc [hunk 3]
```diff
@@ -1799,6 +1799,70 @@ Once "Expand" is selected, the graph is re-drawn to show the children and their
 
 image:expanded-events.png["Expanded Events"]
 
+[[encrypted-provenance]]
+=== Encrypted Provenance Repository
+While OS-level access control can offer some security over the provenance data written to the disk in a repository, there are scenarios where the data may be sensitive, compliance and regulatory requirements, or NiFi is running on hardware not under the direct control of the organization (cloud, etc.). In this case, the provenance repository allows for all data to be encrypted before being persisted to the disk.
```
Comment (andrewmlim) [inline nifi-docs/src/main/asciidoc/user-guide.adoc]: suggest "compliance and regulatory requirements _exist_"

Comment (alopresto) [inline nifi-docs/src/main/asciidoc/user-guide.adoc]: Will do. Thanks.

File: nifi-docs/src/main/asciidoc/user-guide.adoc [hunk 4]
```diff
@@ -1799,6 +1799,70 @@ Once "Expand" is selected, the graph is re-drawn to show the children and their
...
+
+==== How does it work?
+
+The `WriteAheadProvenanceRepository` was introduced in NiFi 1.2.0 and provided a refactored and much faster provenance repository implementation than the previous `PersistentProvenanceRepository`. The encrypted version wraps that implementation with a record writer and reader which encrypt and decrypt the serialized bytes respectively.
+
+The fully qualified class `org.apache.nifi.provenance.EncryptedWriteAheadProvenanceRepository` is specified as the provenance repository implementation in `nifi.properties` as the value of `nifi.provenance.repository.implementation`. In addition, <<administration-guide.adoc#encrypted-write-ahead-provenance-repository-properties,new properties>> must be populated to allow successful initialization.
+
+===== StaticKeyProvider
+The `StaticKeyProvider` implementation defines keys directly in `nifi.properties`. Individual keys are provided in hexadecimal encoding (can also be encrypted like any other sensitive property in `nifi.properties` using the <<administration-guide.adoc#encrypt-config_tool,`./encrypt-config.sh`>> tool in the NiFi Toolkit).
```
Comment (andrewmlim) [inline nifi-docs/src/main/asciidoc/user-guide.adoc]: Suggest break up second sentence into two:  Individual keys are provided in hexadecimal encoding.  The keys can also be encrypted like any other sensitive property in `nifi.properties` using the <<administration-guide.adoc#encrypt-config_tool,`./encrypt-config.sh`>> tool in the NiFi Toolkit.

File: nifi-docs/src/main/asciidoc/user-guide.adoc [hunk 5]
```diff
@@ -1799,6 +1799,70 @@ Once "Expand" is selected, the graph is re-drawn to show the children and their
...
+
+Within the NiFi UI/API, there is no detectable difference between an encrypted and unencrypted provenance repository. The Provenance Query operations work as expected with no change to the process.
+
+==== Potential Issues
+* Switching between unencrypted and encrypted repositories
+** If a user has an existing repository that is not encrypted and switches their configuration to use an encrypted repository, the application writes an error to the log but starts up. However, previous events are not accessible through the provenance query interface and new events will overwrite the existing events. The same behavior occurs if a user switches from an encrypted repository to an unencrypted repository. Automatic roll-over is a future effort (https://issues.apache.org/jira/browse/NIFI-3722[NIFI-3722]) but NiFi is not intended to be long-term storage for provenance events so the impact should be minimal.
```
Comment (andrewmlim) [inline nifi-docs/src/main/asciidoc/user-guide.adoc]: Change to:  Automatic roll-over is a future effort (https://issues.apache.org/jira/browse/NIFI-3722[NIFI-3722]) but NiFi is not intended for long-term storage of provenance events so the impact should be minimal.

File: nifi-docs/src/main/asciidoc/user-guide.adoc [hunk 6]
```diff
@@ -1799,6 +1799,70 @@ Once "Expand" is selected, the graph is re-drawn to show the children and their
...
+*** Logic to handle encrypted -> unencrypted seamlessly as long as the key provider available still has the keys used to encrypt the events (see **Key Rotation**)
+*** Logic to handle unencrypted -> encrypted seamlessly as the previously recorded events simply need to be read with a plaintext schema record reader and then written back with the encrypted record writer
```
Comment (andrewmlim) [inline nifi-docs/src/main/asciidoc/user-guide.adoc]: Feel like these two sub-bullets need some introductory text to make a smooth from the previous section.

File: nifi-docs/src/main/asciidoc/user-guide.adoc (hunk 4 above)
Comment (alopresto) [inline nifi-docs/src/main/asciidoc/user-guide.adoc]: Will do. Thanks.

File: nifi-docs/src/main/asciidoc/user-guide.adoc (hunk 5 above)
Comment (alopresto) [inline nifi-docs/src/main/asciidoc/user-guide.adoc]: Will do. Thanks.

File: nifi-docs/src/main/asciidoc/user-guide.adoc (hunk 6 above)
Comment (alopresto) [inline nifi-docs/src/main/asciidoc/user-guide.adoc]: Will do. Thanks.

**Comment by bbende** [comment]
+1 on this, looks good!

**Comment by alopresto** [comment]
Thanks Bryan. Merging.

**Pull request closed without merging**